server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,433 @@
+        String summary = "No master has been observed recently";
+        Map<String, Object> details = new HashMap<>();
+        Collection<HealthIndicatorImpact> impacts = getUnstableMasterImpacts();
+        List<UserAction> userActions = getContactSupportUserActions(explain);
+        return createIndicator(
+            stableMasterStatus,
+            summary,
+            explain ? new SimpleHealthIndicatorDetails(details) : HealthIndicatorDetails.EMPTY,
+            impacts,
+            userActions
+        );
+    }
+
+    /**
+     * This returns true if this node has seen a master node within the last few seconds
+     * @return true if this node has seen a master node within the last few seconds, false otherwise
+     */
+    private boolean hasSeenMasterInVeryRecentPast() {
+        // If there is currently a master, there's no point in looking at the history:
+        if (clusterService.state().nodes().getMasterNode() != null) {
+            return true;
+        }
+        return masterHistoryService.getLocalMasterHistory().hasSeenMasterInLastNSeconds((int) veryRecentPast.seconds());
+    }
+
+    /*
+     * If we detect that the master has gone null 3 or more times (by default), we ask the MasterHistoryService to fetch the master
+     * history as seen from the most recent master node so that it is ready in case a health API request comes in. The request to the
+     * MasterHistoryService is made asynchronously, and populates the value that MasterHistoryService.getRemoteMasterHistory() will return.
+     * The remote master history is ordinarily returned very quickly if it is going to be returned, so the odds are very good it will be
+     * in place by the time a request for it comes in. If not, this indicator will briefly switch to yellow.
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        DiscoveryNode currentMaster = event.state().nodes().getMasterNode();
+        DiscoveryNode previousMaster = event.previousState().nodes().getMasterNode();
+        if (currentMaster == null && previousMaster != null) {
+            if (masterHistoryService.getLocalMasterHistory().hasMasterGoneNullAtLeastNTimes(acceptableNullTransitions + 1)) {
+                DiscoveryNode master = masterHistoryService.getLocalMasterHistory().getMostRecentNonNullMaster();
+                /*
+                 * If the most recent master was this box, there is no point in making a transport request -- we already know what this
+                 * box's view of the master history is
+                 */
+                if (master != null && clusterService.localNode().equals(master) == false) {
+                    masterHistoryService.refreshRemoteMasterHistory(master);
+                }
+            }
+        }
+    }
+
+    /**
+     * XContentBuilder doesn't deal well with ToXContentFragments (which is what DiscoveryNodes are). Also XContentBuilder doesn't do well
```

Comment:
Can you maybe elaborate what doesn't work with `ToXContentFragments` when it comes to nulls ?

Maybe I'm missing something but wouldn't we handle the nulls when printing the `details` map? I don't think we should be needing another entity to wrap the `DiscoveryNode` as we don't print it directly but via a map that's in our control (the `details` map)

Reply:
I can remove this class and inline the logic.

Reply:
Oh, the reason is that this class does not actually control the printing -- SimpleHealthIndicatorDetails does. I'm not sure how best to handle it in there. We don't want to inspect the type in there and do special handling for different types (I don't think). So I have to change it from DiscoveryNode to something that won't fail. I think the options are to either convert it to a non-fragment ToXContentObject (which is what is being done now) or maybe just convert it to a String? What do you think?

Reply:
Or another option is that I could move this logic into the toXContent method of SimpleHealthIndicatorDetails for any ToXContent value where isFragment() is true

Reply:
Could we take full control over the representation of the details (and avoid the `SimpleHealthIndicatorDetails` altogether) like we do in the `InstanceHasMasterIndicator`? https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/cluster/coordination/InstanceHasMasterHealthIndicatorService.java#L73

Reply:
OK I switched over to custom HealthIndicatorDetails in this class (not counting the one for the "haven't seen a master recently" branch that is just a stub in this PR), and I've added several unit tests around it. It seems to be working well.